### ISSUE TITLE
Inconsistent Blocking Connection hanging when consumer is idle

### ISSUE THREAD

**Issue by sw360cab** [pika/pika#1046]
Pika: 0.12.0b2
OS: Ubuntu

When the connection is left idle for a long time (e.g. 10 hours). A consumer hangs and does not consume anymore messages. 
This used to happen on pika 0.10 and it appeared to be solved in v. 0.11
Following example in [Pika Examples](https://github.com/pika/pika/blob/master/examples/blocking_consume_recover_multiple_hosts.py), the issue could be reproduced.

Moreover:
- the same problem is difficult to reproduce when pushing to localhost (respect to RabbitMQ in another host/network)
- no specific output from pika (at debug level)
- RabbitMQ does not simply to complain about heartebeats
- When pushing new messages, it appears a "strange log", that looks like it comes from pika application (that should consume)
```
rabbit                   | =WARNING REPORT==== 15-May-2018::08:59:41 ===
rabbit                   | closing AMQP connection <0.1090.0> (192.168.1.1:44386 -> 172.18.0.2:5672):
rabbit                   | client unexpectedly closed TCP connection
```

Apart from big headaches, I would like to point out that in pika 0.11.2 this problem should not be present, but another takes places. When operating very long tasks connection is dropped as well.

**Comment by sw360cab** [comment]
Dear @lukebakken thank you for your comment.
I supposed (since I have used for years) that Pika has to be considered the "official" (state-of-the-art) RabbitMQ Python client.

BTW I would be happy to know in which environment you feel confident your [Consumer Example](https://github.com/pika/pika/blob/master/examples/blocking_consume_recover_multiple_hosts.py), works with the condition I described.
Otherwise we can state that the issue is simply there whatever environment I run pika.

**Comment by lukebakken** [comment]
> I supposed (since I have used for years) that Pika has to be considered the "official" (state-of-the-art) RabbitMQ Python client

Pika is the Python AMQP client that is maintained, in part, by RabbitMQ team members, but it is still officially a community project.

> BTW I would be happy to know in which environment you feel confident your Consumer Example, works with the condition I described.

As far as I know, this is the first time this issue (long-idle connection) has been reported. If you can provide the information I requested in [this comment](https://github.com/pika/pika/issues/1046#issuecomment-389206702), I can try to reproduce the issue.

Since you say there is nothing logged by Pika at the debug level, it may be most helpful for you to reproduce the issue while capturing TCP traffic to/from port `5672` with Wireshark. That will show exactly what is happening.

**Comment by lukebakken** [comment]
It's worth a try to use `SelectConnection` but really `BlockingConnection` should work fine - it should continue to send heartbeats and not close the connection as you see in the logs. In your environment, I assume that your Pika program is running on one server, RabbitMQ on another, and there is *no* firewall involved.

**Comment by sw360cab** [comment]
I will give a try to 

> SelectConnection

No, my environment is on a server running Pika, contacting another server running RabbitMQ.
When I start the program it works smootly, then after a while (suppose an hour) if I try to publish message to RabbitMQ and it works (test app communicating and publishing on rabbit queue), instead the consumer app (as the one in the example) on the remote server is still up and idle (the process is still working), but it is simply hanging.

**Comment by michaelklishin** [comment]
What exactly does "simply hanging" mean? Are there no deliveries? Have you checked server logs to verify that there are no [client connection loss events](http://www.rabbitmq.com/logging.html#connection-lifecycle-events) or [missed heartbeats](http://www.rabbitmq.com/heartbeats.html)?

**Comment by levsa** [comment]
> Apart from big headaches, I would like to point out that in pika 0.11.2 this problem should not be present, but another takes places. When operating very long tasks connection is dropped as well.

Is this due to that heartbeat response times out since the thread is occupied handling the previous message? That happened to me and that is why I turned of heartbeats.

**Comment by levsa** [comment]
I'm seeing something similar. I'm running without heartbeats and after a few hours of inactivity, messages end up in unacked state without returning from the `channel.consume()` call I'm using (see code below). After I restart the process and connection is made, consuming proceeds as normal.

My listen-loop looks as follows:

```
        connection = None
        try:
            connection = self._connection_factory.create()  # BlockingConnection
            channel = connection.channel()
            channel.queue_declare('incoming', durable=True)
            logging.info("Waiting for RabbitMQ messages")
            generator = channel.consume('incoming', inactivity_timeout=60)
            for message_tuple in generator:
                if message_tuple is not None:  # None returned after inactivity timeout
                    # tuple(spec.Basic.Deliver, spec.BasicProperties, str or unicode)
                    method, properties, body = message_tuple
                    if method is not None and body is not None:
                        self._handle(channel, method, properties, body)
            channel.cancel()
        except Exception:
            logging.exception("Exception in message receiver loop")
        if connection is not None:
            connection.close()
	
```

The loop seems to hang, i.e., I cannot see any logs printing "Waiting for..." in the loop above, even though there should be with an interval of 60s.

**Comment by levsa** [comment]
Actually, I don't see that `channel.consume()` returns at all anymore. The debug logs I see are:

```
[I 2018-05-18 16:00:19,309] blocking_connection:1911: Created new queue consumer generator <_QueueConsumerGeneratorInfo params=('incoming', False, False) consumer_tag='ctag1.07c4c99a93494edcbb3dfa36e7e9dc54'>
[D 2018-05-18 16:00:19,309] select_connection:203: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x10ba024e0> with deadline=1526652079.30992 and callback=<bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10b9957c8>>; now=1526652019.30992; delay=60
[D 2018-05-18 16:01:19,312] select_connection:203: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x10ba02278> with deadline=1526652139.312602 and callback=<bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10b9957c8>>; now=1526652079.312602; delay=60
[D 2018-05-18 16:02:19,317] select_connection:203: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x10ba02b38> with deadline=1526652199.3177202 and callback=<bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10b9957c8>>; now=1526652139.3177202; delay=60
[D 2018-05-18 16:03:19,322] select_connection:203: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x10ba024e0> with deadline=1526652259.322846 and callback=<bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10b9957c8>>; now=1526652199.322846; delay=60
[D 2018-05-18 16:04:19,323] select_connection:203: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x10ba02278> with deadline=1526652319.323476 and callback=<bound method _CallbackResult.signal_once of <pika.adapters.blocking_connection._CallbackResult object at 0x10b9957c8>>; now=1526652259.323476; delay=60
```

i.e. `consume()` method is not returning as expected when `inactivity_timeout` is given.

**Comment by sw360cab** [comment]
I would like to share what is happening to me.
I have an application based on Pika and employing RabbitMq. 
* The application is made of serveral "blocks" in a sort of pipeline. 
* Each block will:
  * consume message from the previous block
  * produce a message to the next block
* the initial block is triggered from an external event, meaning that external applications will publish messages onto the channel/queue where the initial block is consuming
* each block will be spawned in a thread, but everything regarding Pika connection for the same block will stay in the same thread
* each block will consume a message from a queue and produce message to another (shared with the following block in the pipeline)
* Heartbeats have been disabled as they usually drop connections

This application has already been in production (using Pika 0.10) for at least 2 years, having no issue on Pika. Issue started when application and RabbitMQ server (v. 3.6.6) have to be deployed on different networks / servers. As long as they remained on the same machine (precisely a docker network), the issue was not there.

The relevant part of the application (that consumes messages) has been coded exactly as this repo example:

```` 
def configure(self):
    exchange_name = exchanger['exchange_name']
    hostname = self.read_config_value('host','localhost')
    username = self.read_config_value('username','guest')
    passwd =  self.read_config_value('passwd','guest')
    try:
      # Basic Connection configurtion
      credentials = pika.PlainCredentials(username, passwd)
      connParams = pika.ConnectionParameters(host=str(hostname), credentials=credentials, 
      heartbeat_interval=0, connection_attempts=5, retry_delay=1)
      connection = pika.BlockingConnection(connParams)
      channel = connection.channel()
      channel.exchange_declare(exchange=exchange_name,
                               exchange_type='direct')
      return connection, channel
    except pika.exceptions.AMQPConnectionError:
      self.logger.warn("RabbitMQ may have not been started...")
      import sys
      sys.exit(1)

  def bind(self, channel):
    """
      Binding
    """
    # Named Queue
    if self.hasNamedQueue():
      channel.basic_qos(prefetch_count=1)
      result = channel.queue_declare(queue=self.named_queue_name, durable=True)
    else:
      result = channel.queue_declare(exclusive=True)
    queue_name = result.method.queue
    routing_key = self.inputTier['routing_key']
    channel.queue_bind(exchange=self.inputTier['exchange_name'],
                        queue=queue_name,
                        routing_key = routing_key)
    return queue_name

  def consume(self):
    """
      Consuming incoming messages.
      Emit errors to failureWorker.
    """
    def processMsg(ch, method, properties, body):
      self.logger.debug("[x] %r:%r" % (method.routing_key, body,))
      ch.basic_ack(delivery_tag=method.delivery_tag)
    
    while(True):
      try:
        # Connect & Bind Queue
        connection, channel = self.configure()
        queue_name = self.bind(channel)

        channel.basic_consume(processMsg,
                queue=queue_name,
                no_ack=False)
        channel.start_consuming()
      # Do not recover if connection was closed by broker
      except pika.exceptions.ConnectionClosed:
        break
      # Do not recover on channel errors
      except pika.exceptions.AMQPChannelError:
        break
      except KeyboardInterrupt:
        channel.stop_consuming()
        break
      # Recover on all other connection errors
      except pika.exceptions.AMQPConnectionError:
        continue
    connection.close()
````

Here is how I could reproduce the issue:
1. Start RabbitMQ and application (@ 5pm, timing is relevant)
2. Run test script (that publish messages). The application works smootly and can do is job as well as RabbitMQ.
3. Run the test again and again and no issue at all
4. Going home resting from headaches
5. Getting back in the morning and audit logs. The issue occurred. 

See what i found the following day (I am using a "pipeline" of 2 "blocks" respect to what I described above):
* at midnight (I suppose that) blocks simply reconnected to RabbitMQ:
RabbitMQ:
````
=INFO REPORT==== 16-May-2018::00:00:01 ===
accepting AMQP connection <0.2383.0> (192.168.1.1:44530 -> 192.168.1.100:5672)

=INFO REPORT==== 16-May-2018::00:00:01 ===
accepting AMQP connection <0.2386.0> (192.168.1.1:44532 -> 192.168.1.100:5672)
````

PIka:
````
2018-05-16 00:00:01 DEBUG - pika.callback: Calling <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('192.168.1.226', 44530)->('130.0.131.174', 5672) params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>> for "1:Basic.ConsumeOk"
2018-05-16 00:00:01 DEBUG - pika.callback: Calling <bound method Channel._on_synchronous_complete of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('192.168.1.226', 44534)->('130.0.131.174', 5672) params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>> for "1:Basic.ConsumeOk"
2018-05-16 00:00:01 DEBUG - pika.channel: 0 blocked frames
2018-05-16 00:00:01 DEBUG - pika.channel: 0 blocked frames
2018-05-16 00:00:01 DEBUG - pika.callback: Calling <bound method Channel._on_eventok of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('192.168.1.226', 44530)->('130.0.131.174', 5672) params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>> for "1:Basic.ConsumeOk"
2018-05-16 00:00:01 DEBUG - pika.callback: Calling <bound method Channel._on_eventok of <Channel number=1 OPEN conn=<SelectConnection OPEN socket=('192.168.1.226', 44534)->('130.0.131.174', 5672) params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>> for "1:Basic.ConsumeOk"
2018-05-16 00:00:01 DEBUG - pika.channel: Discarding frame <METHOD(['channel_number=1', 'frame_type=1', "method=<Basic.ConsumeOk(['consumer_tag=ctag1.a50a662ba47944929f287e448ebab64a'])>"])>
2018-05-16 00:00:01 DEBUG - pika.channel: Discarding frame <METHOD(['channel_number=1', 'frame_type=1', "method=<Basic.ConsumeOk(['consumer_tag=ctag1.057fa9bff5484e9ab0ceac20051a930a'])>"])>
````

* @ 10 am the day after, I run again the test, which publish a message:

RabbitMQ:
````
=INFO REPORT==== 16-May-2018::10:04:18 ===
accepting AMQP connection <0.6235.0> (192.168.1.1:49338 -> 192.168.1.100:5672)

=WARNING REPORT==== 16-May-2018::10:04:18 ===
closing AMQP connection <0.238.0> (192.168.1.1:49338 -> 192.168.1.100:5672):
connection_closed_abruptly
````
for the message published by the test.

But the application does not start consuming, moreover Pika is not printing anything.
As long as OS (Ubuntu) is saying the application process is still UP.
Here is the strange, RabbitMQ logs:
````
=WARNING REPORT==== 16-May-2018::10:04:18 ===
closing AMQP connection <0.2383.0> (192.168.1.1:44530 -> 192.168.1.100:5672):
connection_closed_abruptly

=WARNING REPORT==== 16-May-2018::10:04:18 ===
closing AMQP connection <0.2389.0> (192.168.1.1:44534 -> 192.168.1.100:5672):
connection_closed_abruptly
````

BTW the application would stop or exit, but from the Pika point of view it would simply stay and hang.
Last but not least if I force kill application and restart it, It consumes the message published by the test and never consumed before.

I can state that the same happens if I wait a couple hours. For sake of simplicity I tried waiting longer, so all night.

I just want to point out again that as today this application and its architecture are already in production, and this kind of issue does not appear. They do appear where I move RabbitMQ to an address different than "localhost".
Basically I work with Docker. I made these attempt in a "native" environment to avoid that something was related to Docker. As the issue occurred the same I suppose no.

I would be glad if I can cooperate to find and solve the issue. Or if you have any suggestion to replace `Blocking Connection` with something more reliable. Also heartbeats are a long chapter, at the end disabling them has always solved troubles but I can be work or using a naïf approach.

This application is a crucial piece in our company and in this unstable situation sooner or later I would have to drop Pika and RabbitMQ in flavour of something else, and this would be sad and disappointing. Not only for the need of rewriting classes of code.

**Comment by vitaly-krugl** [comment]
@sw360cab:

Regarding your application that is already in Production: in your production environment, is your pika application on the same server as the RabbitMQ broker?

Here is some additional food for thought: when something goes wrong on the network that severs the TCP connection abruptly such that FIN is not received by the consumer, the consuming end running pika (with heartbeats disabled) has no way to find out that the connection was disconnected. I have seen this happen, for example, when the server dies suddenly; also have seen this when some load balancer drops the connection after a period of inactivity.

If the hang is really due to the lack of incoming FIN, then it would explain the appearance of a hang in BlockingConnection. It would also explain the lack of anything interesting in Pika DEBUG logs. If FIN didn't come in and no pika messages came in, then there would be no activity to log in Pika.

Hence, if a connection goes dead without receiving FIN (or RST), there is no way for an application to find out. Unless, of course, you enable AMQP heartbeat! :).

Since your sample code that reproduces the issue simply discards the messages without taking a long time to process them, go ahead and enable the heartbeat. You might see one of these new outcomes: 1. the connection no longer drops due to heartbeat traffic across some network element that used to sever the connection (not gracefully) when the connection became idle; or 2. BlockingConnection with heartbeat enabled discovers the dropped connection (no heartbeat or other traffic from RabbitMQ) and raises the expected exception to signal loss of connection.

Finally, the reason that your app disables heartbeats is most likely because the long message processing time was causing heartbeat timeout and loss of connection. BlockingConnection in Pika v0.12.0 has a new feature to make it practical to defer processing of messages to another thread, while the BlockingConnection's thread continues the the consumer event pump; when the other thread is done with the message, it can use BlockingConnection's new method `add_callback_threadsafe()` to request that an ACK be sent in a thread-safe way. This feature was back-ported from the Pika 1.0 pre-release work and has an example in the [README.rst](https://github.com/pika/pika/blob/master/README.rst). Look for `add_callback_threadsafe` in this README. This way, lengthy processing of incoming messages can coexist elegantly/efficiently with AMQP heartbeats, and the heartbeats will help detect adverse (not graceful) loss of TCP stream.

**Comment by vitaly-krugl** [comment]
@sw360cab, does this help?

**Comment by sw360cab** [comment]
@vitaly-krugl

Thank you for raising so many arguments, they are all strongly related to my scenario description.

>Regarding your application that is already in Production: in your production environment, is your pika application on the same server as the RabbitMQ broker?

That's something I want to point.
In production application and RabbitMQ live on the some host.
The issues started occurring when RabbitMQ and application are living on different hosts, even if on the same or on different networks.

> Since your sample code that reproduces the issue simply discards the messages without taking a long time to process them, go ahead and enable the heartbeat. You might see one of these new outcomes: 1. the connection no longer drops due to heartbeat traffic across some network element that used to sever the connection (not gracefully) when the connection became idle for some period of time; or 2. BlockingConnection with heartbeat enabled discovers the severed connection (no heartbeat or other traffic from RabbitMQ) and raises the expected exception to signal loss of connection.

About this part (and I thank you for the full and deep explaination), I used to disable heartbeats because they gave extra pain to my application. And since disabling them allowed the application to work (in local and production environment, on a single host), I too naively reached to the conclusion that application can stand with them disabled.

But once I encountered the issue of course I started to attempt dealing with heartbeats. At first I didn't have an exception handler as satisfing as I suppose to have now, so I didn't go in deep for this path. On the other hand Pika 0.11.2 seems to have solved perfectly any issue, but as my application may perform very long tasks (I skipped this partein the snippet for sake of smooth explaination), I encountered many problems in "acking" back messages on "dead" channel connection (as you figured out correctly).
Strangely although pika 0.11.2 seamed to deal perfectly with connection status, it didn't with acking messages after very long processing. Whereas Pika 0.12pre looks like doing exactly the opposite, it dealed the problem of "long task", but fails connection soon.

Eventually I felt myself resting in a "cul-de-sac" / dead end...

Now after your suggestion, I am working on 3 different attempts to allow us investigating more in deep into the problem:

#### attempt # 1

I will enable heartbeats and see whether exception are raised by Pika and handled correctly by the application

#### attempt # 2

I will try to move the callback "ack" part into a `add_callback_threadsafe()` block, eventually re-switching back to Pika 0.11.2

#### attempt # 3 

I will rewrite part of the application to work with `SelectConncetion` with is a little more complex to be handled but probably more reliable and promising

---

As long as I end up with these attempts I will be happy to report to you and see what happened. 
Hope that by the end of the week I can share any result.

**Comment by sw360cab** [comment]
A first attempt adding an heatbeat timeout of 600s reproduces the issue.
I pasted Pika log here:

[https://pastebin.com/K4cJwD3w](https://pastebin.com/K4cJwD3w)

There could be some extra log.... But the attempt was, at 10.40 a message is received and processed, the same message is sent 4 hours later and never received.

Could it be related with previous message handling? I remeber that when I attempted to reproduce the issue earlier I used to:
1. run the application 
2. produce & process several messages
3. wait all night

And then the day after producing further message would leave the application in hanging state.

In this case I ran the application, but I did not produce any message, I just left it up all night.
Starting producing messages the day after works. Waiting 3 hours and producing further messages seem to hang the whole connection (as stated before).

**Comment by lukebakken** [comment]
@sw360cab could you please reproduce this while capturing traffic to port `5672` on your server? You can run the packet capture on the server. This will help us out to diagnose the issue. I will see if I can reproduce it using my home network.

> This application is a crucial piece in our company and in this unstable situation sooner or later I would have to drop Pika and RabbitMQ in flavour of something else, and this would be sad and disappointing. Not only for the need of rewriting classes of code.

Can you downgrade Pika to `0.10` until we figure this out? That would fix your issue, correct?

**Comment by sw360cab** [comment]
> @sw360cab could you please reproduce this while capturing traffic to port 5672 on your server? You can run the packet capture on the server. This will help us out to diagnose the issue. I will see if I can reproduce it using my home network.

Yes I'll try

> Can you downgrade Pika to 0.10 until we figure this out? That would fix your issue, correct?

I am doing already it, as long as I stuck with Pika 0.10 and with  RabbitMQ and application on the same machine, the issue is not occuring.

@lukebakken  Thank you for supporting.

**Comment by lukebakken** [comment]
> I am doing already it, as long as I stuck with Pika 0.10 and with RabbitMQ and application on the same machine, the issue is not occuring.

Just to clarify, if you use Pika `0.10` and RabbitMQ on **different** machines, you also do **not** see the issue. Is that correct?

**Comment by sw360cab** [comment]
@lukebakken I think the issue is there even with 0.10, or similar issue was there.
The motivation of moving to new version of Pika were issues on 0.10 version while changing the architecture of application (rabbit and app on different hosts).
Since I have attempted hundreds of combinations of configurations I don't remeber exactly.
I can try again this scenario too.

**Comment by vitaly-krugl** [comment]
@lukebakken, would you mind including the pika version number in the log content when opening a connection? This would help get our bearings when examining logs. Thx!

**Comment by vitaly-krugl** [comment]
cc @lukebakken 
@sw360cab - regarding:
> A first attempt adding an heatbeat timeout of 600s reproduces the issue. I pasted Pika log here: https://pastebin.com/K4cJwD3w

Which version of Pika are you using in this case?

A couple of strange things in this log:
1. The application closes Blocking connection (including the socket)  at 10:40:04, but the log shows incoming and outgoing heartbeats all the way until 14:20:04.  How could that be? Is there another connection in play here? Was the process forked? Very strange!: ```
2018-05-28 10:40:04 INFO - pika.connection: Disconnected from RabbitMQ at peer.wim.tv:5672 (200): Normal shutdown
2018-05-28 10:40:04 DEBUG - pika.heartbeat: Removing timeout for next heartbeat interval
2018-05-28 10:40:04 DEBUG - pika.adapters.select_connection: remove_timeout: removing timeout <pika.adapters.select_connection._Timeout object at 0x7f69580ef310> with deadline=1527497404.504849 and callback=<bound method HeartbeatChecker.send_and_check of <pika.heartbeat.HeartbeatChecker object at 0x7f69580ef050>>
2018-05-28 10:40:04 DEBUG - pika.callback: Processing 0:_on_connection_closed
2018-05-28 10:40:04 DEBUG - pika.callback: Calling <bound method _CallbackResult.set_value_once of <pika.adapters.blocking_connection._CallbackResult object at 0x7f695aafb830>> for "0:_on_connection_closed"
2018-05-28 10:40:04 DEBUG - pika.callback: Added: {'callback': <bound method SelectConnection._on_connection_start of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 1}
2018-05-28 10:40:04 DEBUG - pika.callback: Incremented callback reference counter: {'callback': <bound method SelectConnection._on_connection_close of <SelectConnection CLOSED socket=None params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>>, 'only': None, 'one_shot': True, 'arguments': None, 'calls': 2}
2018-05-28 10:40:04 INFO - pika.adapters.blocking_connection: Connection closed; result=BlockingConnection__OnClosedArgs(connection=<SelectConnection CLOSED socket=None params=<ConnectionParameters host=peer.wim.tv port=5672 virtual_host=/ ssl=False>>, reason_code=200, reason_text='Normal shutdown')
2018-05-28 10:40:04 DEBUG - pika.heartbeat: Received 305 heartbeat frames, sent 305
2018-05-28 10:40:04 DEBUG - pika.heartbeat: Sending heartbeat frame
2018-05-28 10:40:04 DEBUG - pika.adapters.select_connection: call_later: added timeout <pika.adapters.select_connection._Timeout object at 0x7f69617e4110> with deadline=1527497404.518461 and callback=<bound method HeartbeatChecker.send_and_check of <pika.heartbeat.HeartbeatChecker object at 0x7f69617e4710>>; now=1527496804.52; delay=600
2018-05-28 10:43:39 DEBUG - pika.heartbeat: Received heartbeat frame
2018-05-28 10:48:39 DEBUG - pika.heartbeat: Received heartbeat frame
2018-05-28 10:48:39 DEBUG - pika.heartbeat: Received heartbeat frame
2018-05-28 10:48:39 DEBUG - pika.heartbeat: Received 307 heartbeat frames, sent 306
......
```

**Comment by vitaly-krugl** [comment]
See also #1055

**Comment by lukebakken** [comment]
As expected, I can't reproduce this issue when heartbeats are used. I tested with Pika `0.11.2` and RabbitMQ `3.7.5` with a simple script that sets up a `BlockingConnection` and a consumer. I published several messages and then let the application sit idle for several hours. Debug logs, as expected, showed heartbeats being sent. Subsequent publishes after the "idle" time succeeded.

I'll test this again using Pika `0.12.0.b3` with disabled heartbeats (which works correctly in that version, see #1014).

**Comment by sw360cab** [comment]
@lukebakken 

I am doing my tests with 0.12.0.b2 and heartbeats at 600s

While testing I am capturing packets (Wireshark). Everything is working smoothly since this morning in the scenario.

I will see if by tomorrow the connection is still available leaving it idle all night long.
If this works I would repeat the tests with:
- pika 0.12.0.b3
- dockerized version of RabbitMq

**Comment by lukebakken** [comment]
I have had a test running for 8 hours with the attached code, using Pika `0.12.0b3` with heartbeats disabled. RabbitMQ is running on a separate machine from the test program. I just published 5 messages which were consumed as expected.

[consume.py.txt](https://github.com/pika/pika/files/2074629/consume.py.txt)

I will leave this running all night and will publish messages in the morning.

**Comment by lukebakken** [comment]
I had no problems after leaving the test program running all night. Publishing several messages this morning worked as expected. You can see in the packet capture where TCP keepalives were sent at the default 2 hour interval overnight.

Going forward, please be sure to run Pika in your environment with heartbeats enabled. Version `0.12` and `1.0.0` support the `add_callback_threadsafe` method on `BlockingConnection` that can be used to schedule an acknowledgment from another thread.

Or, a better solution would be to use `SelectConnection` and an async producer/consumer as shown in the examples. Threads that do work in your application can use the `add_timeout` method to schedule an callback to do the acknowledgement on the correct thread.

@sw360cab if you wouldn't mind reporting what you find I would appreciate it. I'm going to close this issue but will keep an eye out for your response and I will re-open this issue if necessary.

[gh-1046-capture-5672.pcapng.gz](https://github.com/pika/pika/files/2076444/gh-1046-capture-5672.pcapng.gz)

**Comment by vitaly-krugl** [comment]
@lukebakken, I think the suggestion to use SelectConnection wouldn't improve the situation, just make his code more complicated due to "callback hell" :). Also, the suggestion to use `add_timeout` to delegate processing to the correct thread looks like a typo. If someone is using SelectConnection directly, and need to schedule a callback inside SelectConnection's I/O loop safely, they should call the `add_callback_threadsafe()` method of that SelectConnection's I/O loop directly (not `add_timeout` which isn't thread-safe)

**Comment by lukebakken** [comment]
Thanks @vitaly-krugl for some reason I thought `add_timeout` was the equivalent of `BlockingConnection.add_callback_threadsafe`

**Comment by vitaly-krugl** [comment]
@lukebakken: `add_timeout()` is the non-thread-safe timer API in 0.x.x
(renamed to `call_later()` in master now because add_timeout is intended to
be used with absolute time while ours was accepting relative)

Technically, select_connection.IOLoop defines the alias `add_callback` that
maps to `select_connection.IOLoop.add_callback_threadsafe`. This was done
because this IOLoop now shares some common Pika utilities with tornado's
`IOLoop` class. select_connection's `IOLoop` API was historically modeled
on tornado's `IOLoop` which has `add_callback()`. But since we're talking
about using SelectConnection's IOLoop here, `add_callback_threadsafe()` is
more self-documenting, so I'd rather advertise it.

**Comment by sw360cab** [comment]
@lukebakken and @vitaly-krugl thank you for your support. I have changed my code (as I stated in my previous comment) and put it in deep testing employing hearthbeats and Wireshark to inspect packets flowing in the net and I definitely upgraded Pika to version 0.12b2.

I don't want to scream it loud yet, but after a week of testing it sems to be working back smootly even with the production scenario. I hope it will keep going like that 👯 

I would like to add more:
- I usually develop on Mac Os (10.12 + Python 2.7), and now I deploy on Ubuntu (right now, 18.04).
- heartbeating of Pika 0.12b2 is working perfectly on MacOs, whereas on Ubuntu it sends heartbeat every heartbeat/2 seconds (e.g heartbeat=600, will send heartbeats every 300 seconds). May this be related to #1055 ?
- I forgot to mention that I was using (also at the time I opened the issue) in the production stage Ubuntu 18.04. Is it possible that I opted for this version (18) of Ubuntu too early? 
I state that because after upgrading system and reinstalling Pika (0.12b2), it seems definitely stable. I haven't had any specific edit on my code since that big ones done before my last comment to the issue.
But some small problems I was encountering disappered simply by upgrading. 
What is your feeling on that?
- I will plan to pass to Pika 0.12b4 or any other stable one in the near future. I will stay tuned on the follow ups of Pika.

Finally thank you for supporting and for being patience and trying the issue environment.
@lukebakken I hope I can keep the issue close 😸

**Comment by lukebakken** [comment]
@sw360cab thank you very much for following up -

> heartbeating of Pika 0.12b2 is working perfectly on MacOs, whereas on Ubuntu it sends heartbeat every heartbeat/2 seconds (e.g heartbeat=600, will send heartbeats every 300 seconds). May this be related to #1055 

Yes, Pika will send heartbeats using an interval that is calculated by taking the negotiated connection timeout value (600 seconds in your case) and dividing it by 2. This will happen no matter the operating system. This change is to make Pika's behavior the same as the RabbitMQ Java client as well as Bunny (the Ruby client).

> I forgot to mention that I was using (also at the time I opened the issue) in the production stage Ubuntu 18.04. Is it possible that I opted for this version (18) of Ubuntu too early?

I doubt it as Python is Python ... I haven't seen reports of distribution-specific issues. I suppose it's possible but it seems that your issues were probably caused by a bug in Pika or your code.

I just released Pika `0.12` yesterday - give it a try! Thank you again for keeping us informed.

https://groups.google.com/forum/#!topic/pika-python/2pyZG-cB5_g
